assayist/client/query.py

```diff
@@ -44,3 +45,134 @@ def get_container_content_sources(build_id):
         'internal_urls': internal_urls,
         'upstream_urls': upstream_urls
     }
+
+
+def get_source_components_for_build(build_id):
+    """
+    Get source code components used in a build.
+
+    In more detail:
+    Get the canonical names of
+    the sources ultimately used for
+    artifacts ultimately embedded in
+    the artifact from a build.
+
+    For example: for a container image build, get the names of:
+    * Upstream projects for RPMs installed in the images
+    * Go packages used to build executables in the images
+    etc
+
+    The canonical names include versions where available.
+
+    This function is implemented using two DB queries. The first
+    establishes the artifacts our query relates to, as well any
+    relationships between them. The second finds the information
+    needed to establish source code versions relating to those
+    artifacts.
+
+    :param int build_id: the Koji build's ID
+    :return: a list of canonical names (package URLs)
+    :rtype: list
+    """
+    build = Build.nodes.get_or_none(id_=str(build_id))
+    if not build:
+        raise NotFound('The requested build does not exist in the database')
+
+    query = """
+    // Find artifacts which (or artifacts which embed artifacts which)...
+    // Note: "*0.." means zero or more edges; if zero edges, 'a' is the
+    //
+    // artifact directly produced by the build.
+    MATCH (a:Artifact) <-[e:EMBEDS*0..]- (:Artifact)
+
+    // Were produced by the build
+            <-[:PRODUCED]- (:Build {{id: {0} }})
+
+    // Return the artifacts and relationships
+    RETURN a, e
+    """.format(repr(str(build_id)))
+
+    results, _ = neomodel.db.cypher_query(query)
+    artifacts_by_id = {}
+    for artifact in [Artifact.inflate(a) for a, _ in results]:
+        artifacts_by_id[(artifact.type_, artifact.archive_id)] = {
+            "artifact": artifact,
+        }
+
+    embedded_artifacts = set()  # needed when we build a tree later
+    for _, edges in results:
+        for edge in edges:
+            embeds, embedded = [Artifact.inflate(node)
+                                for node in edge.nodes]
+            by_id = artifacts_by_id[(embeds.type_, embeds.archive_id)]
+            embeds_list = by_id.setdefault("embeds_ids", set())
+            index = (embedded.type_, embedded.archive_id)
+            embeds_list.add(index)
+            embedded_artifacts.add(index)
+
+    query = """
+    // Find the artifacts
+    // TODO: archive_id is not unique. We should take the
```

Comment:
Are you planning to address this before merging the PR?

Reply:
Nice, thanks. Fixed.